firedrake/dmplex.pyx

```diff
@@ -1014,3 +964,784 @@ def plex_renumbering(PETSc.DM plex, np.ndarray[PetscInt, ndim=1, mode="c"] reord
+    """
+    cdef np.int32_t ri
+
+    # Initiate receiving
+    recv_reqs = []
+    for ri in range(nranks):
+        recv_reqs.append(MPI.comm.Irecv(theirs[offsets[ri] : offsets[ri+1]], ranks[ri]))
+
+    # Initiate sending
+    send_reqs = []
+    for ri in range(nranks):
+        send_reqs.append(MPI.comm.Isend(ours[offsets[ri] : offsets[ri+1]], ranks[ri]))
+
+    # Wait for completion
+    for req in recv_reqs:
+        req.Wait()
+    for req in send_reqs:
+        req.Wait()
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def quadrilateral_facet_orientations(
+    PETSc.DM plex, PETSc.Section vertex_numbering,
+    np.ndarray[np.int32_t, ndim=1, mode="c"] cell_ranks):
+
+    """Returns globally synchronised facet orientations (edge directions)
+    incident to locally owned quadrilateral cells.
+
+    :arg plex: The DMPlex object encapsulating the mesh topology
+    :arg vertex_numbering: Section describing the universal vertex numbering
+    :arg cell_ranks: MPI rank of the owner of each (visible) non-owned cell,
+                     or -1 for (locally) owned cell.
+    """
+    cdef:
+        np.int32_t nranks
+        np.int32_t *ranks = NULL
+        np.int32_t *offsets = NULL
+        np.int32_t *facets = NULL
+        np.int32_t *facet2index = NULL
+
+        PetscInt nfacets, nfacets_shared, fStart, fEnd
+
+        np.ndarray[np.int32_t, ndim=1, mode="c"] affects
+        np.ndarray[np.int32_t, ndim=1, mode="c"] ours, theirs
+        np.int32_t conflict, value, f, i, j
+
+        PetscInt ci, size
+        PetscInt cells[2]
+
+        np.ndarray[np.int8_t, ndim=1, mode="c"] result
+
+    # Get communication lists
+    get_communication_lists(plex, vertex_numbering, cell_ranks,
+                            &nranks, &ranks, &offsets, &facets, &facet2index)
+    nfacets_shared = offsets[nranks]
+
+    # Discover edge direction dependencies in the mesh locally
+    fStart, fEnd = plex.getHeightStratum(1)
+    nfacets = fEnd - fStart
+
+    result = np.full(nfacets, -1, dtype=np.int8)
+    affects = locally_orient_quadrilateral_plex(plex,
+                                                vertex_numbering,
+                                                <np.int32_t *>cell_ranks.data,
+                                                facet2index,
+                                                nfacets_shared,
+                                                <np.int8_t *>result.data)
+    CHKERR(PetscFree(facet2index))
+
+    # Initialise shared edge directions and assign weights
+    ours = MPI.comm.size * np.arange(nfacets_shared, dtype=np.int32) + MPI.comm.rank
+    for i in range(nfacets_shared):
+        if affects[i] != nfacets_shared:
+            j = cabs(affects[i])
+            if cabs(ours[i]) < cabs(ours[j]):
+                if affects[i] >= 0:
+                    ours[i] = ours[j]
+                else:
+                    ours[i] = cneg(ours[j])
+    theirs = np.empty_like(ours)
```

Comment:
OK, so ours is full of the local view of what the orientations are.  And theirs will be filled by the remote orientation view.